### ISSUE TITLE
Uncaught TypeError: this.buildErrs.push is not a function

### ISSUE THREAD

**Issue by HWiese1980** [James-Yu/Atom-LaTeX#110]
[Enter steps to reproduce:]

1. large project consisting of multiple tex files in subfolders
2. build project

**Atom**: 1.19.5 x64
**Electron**: 1.6.9
**OS**: Ubuntu 16.04.3
**Thrown From**: [atom-latex](https://github.com/James-Yu/Atom-LaTeX) package 0.8.1

### Stack Trace

Uncaught TypeError: this.buildErrs.push is not a function

```
At /home/xxx/.atom/packages/atom-latex/lib/builder.coffee:51

TypeError: this.buildErrs.push is not a function
    at Builder.module.exports.Builder.buildProcess (/packages/atom-latex/lib/builder.coffee:51:16)
    at /packages/atom-latex/lib/builder.coffee:76:10)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

### Commands

```
     -0:23.5.0 intentions:highlight (input.hidden-input)
     -0:23.1.0 atom-latex:build (input.hidden-input)
```

### Non-Core Packages

```
atom-latex 0.8.1 
atom-ros 0.1.0 
atom-typescript 11.0.7 
autocomplete-cmake 0.3.1 
build-catkin-tools 1.0.0 
build-cmake 0.10.2 
busy-signal 1.4.3 
docker 0.8.0 
document-outline 1.6.1 
intentions 1.1.5 
language-cmake 0.3.0 
language-docker 1.1.8 
language-latex 1.1.1 
latex 0.46.0 
linter 2.2.0 
linter-docker 0.2.1 
linter-ui-default 1.6.8 
MagicPython 1.0.12 
python-tools 0.6.9 
split-diff 1.5.1 
```

**Comment by ashthespy** [comment]
Any chance you have a small MWE that I could try on my system to reproduce?

**Comment by HWiese1980** [comment]
I'll see if I can come up with some... give me a second...

**Comment by HWiese1980** [comment]
It's reproducable on my system with the most basic of all MWEs:

    \documentclass{scrreprt}
    \begin{document}
    \end{document}

That's it. 

Here's by the way what the JavaScript console in Atom outputs about this error: 

    Download the React DevTools and use an HTTP server (instead of a file: URL) for a better development experience: https://fb.me/react-devtools
    /home/dfki.uni-bremen.de/hwiese/.atom/packages/linter-ui-default/node_modules/react/lib/lowPriority…:40 Warning: Accessing PropTypes via the main React package is deprecated, and will be removed in  React v16.0. Use the latest available v15.* prop-types package from npm instead. For info on usage, compatibility, migration and more, see https://fb.me/prop-types-docs
    printWarning @ /home/dfki.uni-bremen.de/hwiese/.atom/packages/linter-ui-default/node_modules/react/lib/lowPriority…:40
    /home/dfki.uni-bremen.de/hwiese/.atom/packages/linter-ui-default/node_modules/react-dom/lib/ReactDOM.js:73 Download the React DevTools and use an HTTP server (instead of a file: URL) for a better development experience: https://fb.me/react-devtools
    manager.coffee [sm]:197 RootWatcher Init: 0.095ms
    builder.coffee [sm]:51 Uncaught TypeError: this.buildErrs.push is not a function
        at Builder.module.exports.Builder.buildProcess (builder.coffee [sm]:34)
        at ChildProcess.<anonymous> (builder.coffee [sm]:52)
        at emitTwo (events.js:106)
        at ChildProcess.emit (events.js:191)
        at Process.ChildProcess._handle.onexit (internal/child_process.js:215)

**Comment by HWiese1980** [comment]
Just to mention it: I have in the meantime updated my atom installation from 1.18.0 to 1.19.5. That might be the reason for the differing line numbers in the JavaScript output and the initial bug report.

**Comment by HWiese1980** [comment]
This is a very weird error because a line before where the exception occurs there's another call to pretty much the same function which does not crash. 

    @buildLogs.push '' // works
    @buildErrs.push '' // crashes

Although the initialization of the members of the class `Builder` in that coffee script is the same. 

    @buildLogs = []
    @buildErrs = []

I don't see why it should crash there. It seems more like an error caused by some call stack corruption but it's reproducable every time and the error is always the same. That speaks against a call stack corruption. Otherwise I think I would see random errors happening at different points...

It's weird...

**Comment by ashthespy** [comment]
Thanks for the details. Indeed something quite strange. 
Did you had Atom-LaTeX `0.8.1` running on Atom `v1.18.0`? Because the `TextBuffer.save()` changed to an async function from Atom `v1.19.0`.  

Maybe a long shot, but can you try running `apm rebuild-module-cache` ?

**Comment by HWiese1980** [comment]
I've run `apm rebuild-module-cache`, every step came back OK. Unfortunately the error persists. 

I might have had Atom-LaTeX 0.8.1 running on Atom v1.18.0. I cannot tell for sure though. I'll try uninstalling and reinstalling it and see if that helps.

**Comment by HWiese1980** [comment]
Nope... didn't help. The error persists also after install-cycling Atom-LaTeX.

**Comment by HWiese1980** [comment]
Are you investigating the problem, @ashthespy?

**Comment by ashthespy** [comment]
@HWiese1980 I dug a little, but did not find anything useful over and above the standard 'Have you tried turning it off and on again' 😛 

A rather dirty solution that might work is to comment out the `@buildErrs` references in `builder.coffee` That should hopefully trigger rebuilding the package source, which *might* fix your issue. 

Unfortunately, this is rather outside my domain, and all my suggestions are just shots in the dark!

**Comment by HWiese1980** [comment]
Yeah, the dirty trick does in fact work. I'm really wondering why it is, though... or rather why it isn't without.

Anyway... I'd recommend leaving the issue open for the time being. Maybe someone flies by who knows an answer.

**Comment by ashthespy** [comment]
@HWiese1980  Does this error still persist after the update?

**Comment by gwdg-chripo** [comment]
@ashthespy The update to v0.8.2 did not fix this error for me. I am using Atom v1.20.1 on MacOS Sierra and MacTex.

**Comment by ashthespy** [comment]
@gwdg-chripo  Didn't know there were more people facing this issue!
Can you provide a stack trace of the issue?

**Comment by gwdg-chripo** [comment]
@ashthespy Sure, looks pretty much like the original one from @HWiese1980. Please let me know if I can assist you further in finding the cause of this issue.

**Atom**: 1.20.1 x64
**Electron**: 1.6.9
**OS**: Mac OS X 10.12.6
**Thrown From**: [atom-latex](https://github.com/James-Yu/Atom-LaTeX) package 0.8.2


### Stack Trace

Uncaught TypeError: this.buildErrs.push is not a function

```
At /Users/cpohl/.atom/packages/atom-latex/lib/builder.coffee:51

TypeError: this.buildErrs.push is not a function
    at Builder.module.exports.Builder.buildProcess (/packages/atom-latex/lib/builder.coffee:51:16)
    at /packages/atom-latex/lib/builder.coffee:76:10)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

### Commands

```
     -1:16.2.0 application:add-project-folder (ul.list-inline.tab-bar.inset-panel)
     -0:37.5.0 tree-view:show (atom-workspace.workspace.scrollbars-visible-when-scrolling.theme-solarized-light-syntax.theme-solarized-dark-ui)
```

### Non-Core Packages

```
aligner 1.2.4 
aligner-javascript 1.3.0 
aligner-puppet 0.1.1 
atom-clock 0.1.15 
atom-latex 0.8.2 
busy-signal 1.4.3 
file-icons 2.1.12 
intentions 1.1.5 
language-groovy 0.7.0 
language-puppet 0.23.0 
linter 2.2.0 
linter-checkbashisms 0.1.5 
linter-json-lint 0.1.2 
linter-puppet-lint 0.8.3 
linter-ui-default 1.6.10 
minimap 4.29.7 
solarized-dark-ui 0.3.2 
sync-settings 0.8.3 
termination 0.6.1 
```

**Comment by HWiese1980** [comment]
@gwdg-chripo Thanks for taking over. We've had a long weekend here in Germany and I haven't had the time yet to check whether the bug persists for me. I will have today, hopefully, as I've got some LaTeX stuff to do. So I should be able to test it. I hope I'll not forget about it...

**Comment by HWiese1980** [comment]
@ashthespy I've been able to check and I have to say, alas, the bug persists.

**Comment by gwdg-chripo** [comment]
@ashthespy I think I narrowed this issue down a bit. Seems like a bad combination of using `lualatex` with the `-shell-escape` parameter followed by another command in a custom chain is causing the error.

So both following settings in `.latexcfg` will cause the error to appear (e.g. substitute <some-other-cmd> with `echo 'Hello World'`):

`{
  "toolchain" : "lualatex -shell-escape %ARG %DOC && <some-other-cmd>"
}`

`{
  "toolchain" : "<some-other-cmd> && lualatex -shell-escape %ARG %DOC && <some-other-cmd>"
}`

But all of the following .latexcfg settings will work fine:

`{
  "toolchain" : "lualatex %ARG %DOC && <some-other-cmd>"
}`

`{
  "toolchain" : "<some-other-cmd> && lualatex %ARG %DOC && <some-other-cmd>"
}`

`{
  "toolchain" : "lualatex %ARG %DOC && lualatex -shell-escape %ARG %DOC"
}`

`{
  "toolchain" : "pdflatex -shell-escape %ARG %DOC && pdflatex -shell-escape %ARG %DOC && <some-other-cmd>"
}`

So basically after running `lualatex` with the `-shell-escape` parameter somewhere in the custom chain, any further command appended with `&&` will raise the error.

**Comment by ashthespy** [comment]
@gwdg-chripo 
Thanks for the detailed report - this issue was like an itch that I could not scratch. But with these details, I shall look into it. 
I am glad that you have `latekmk` working for you in the meantime.

**Comment by HWiese1980** [comment]
... which still remains really weird from a programmer's point of view because of https://github.com/James-Yu/Atom-LaTeX/issues/110#issuecomment-327486689

**Comment by gwdg-chripo** [comment]
@ashthespy I'm sorry, but your MWE raises the error on my side. Please find attached a tar archive with mini.tex and .latexcfg as used to reproduce the exception here. My atom-latex settings regarding toolchain, compiler and compiler arguments are set to the defaults.

I'm currently using MacTex (all packages up-to-date via TeX Live), but will try the MWE on different OS/TeX distro soon.

@HWiese1980 agreed. :-)

[atom-latex-uncaught-typeerror.tar.gz](https://github.com/James-Yu/Atom-LaTeX/files/1414147/atom-latex-uncaught-typeerror.tar.gz)

**Comment by ashthespy** [comment]
@gwdg-chripo  The mysteries deepen again! I will spin up VM Ubuntu system and check it out over the weekend. (Couldn't get reproduce with MiKTeX on Win10)

**Comment by gwdg-chripo** [comment]
@ashthespy I can confirm that the error does not appear on Windows 10 (using TeXLive here, not MikTeX). It does show up under Linux (Ubuntu 16.04, again with TeXLive). Could the error be related to the underlying shell environment for command execution (e.g. PowerShell on Windows and bash on Linux and MacOS)?

**Comment by ashthespy** [comment]
@gwdg-chripo Thanks for the detailed investigation I have been having a busy few weeks off late.

This is a very silly bug in hindsight. And as your investigations show -I would have never have experienced it on my Win box. Maybe it's time to write better specs/ test more cross platform! 
Have patched it, thanks again for tackling this! :-)
```diff
diff --git a/lib/builder.coffee b/lib/builder.coffee
index 1ff5a3b..3c65de7 100644
--- a/lib/builder.coffee
+++ b/lib/builder.coffee
@@ -63,7 +63,7 @@ class Builder extends Disposable
       @buildLogs[@buildLogs.length - 1] += data
 
     @currentProcess.stderr.on 'data', (data) =>
-      @buildErrs += data
+      @buildErrs[@buildErrs.length - 1] += data
 
     @currentProcess.on 'error', (err) =>
       # Fatal executable error
```

**Comment by gwdg-chripo** [comment]
@ashthespy First of all, thank you for maintainig this package, Atom with atom-latex is my favorite cross-platform LaTeX environment, and I hope it will be around for long. :simple_smile: :thumbsup:

Unfortunately, your patch raises another exception for me, now containing the aforementioned error message with $TERM:

```
tput: No value for $TERM and no -T specified
,
```

The build log shows, that bibtex is executed, although it is not included in the .latexcfg toolchain (I'm using the MWE discussed earlier):

```
> bibtex mini

This is BibTeX, Version 0.99d (TeX Live 2017)
The top-level auxiliary file: mini.aux
I found no \citation commands---while reading file mini.aux
I found no \bibdata command---while reading file mini.aux
I found no \bibstyle command---while reading file mini.aux
(There were 3 error messages)
```
